File: src/main/java/by/dev/madhead/jarvis/creator/ChangeCreator.java
package by.dev.madhead.jarvis.creator;

import by.dev.madhead.jarvis.model.Author;
import by.dev.madhead.jarvis.model.Change;
import by.dev.madhead.jarvis.util.AddressSearcher;
import hudson.FilePath;
import hudson.scm.ChangeLogSet;
import org.eclipse.jgit.lib.ObjectId;
import org.eclipse.jgit.lib.Repository;
import org.eclipse.jgit.lib.RepositoryBuilder;
import org.eclipse.jgit.revwalk.RevCommit;
import org.eclipse.jgit.revwalk.RevWalk;

import java.io.File;
import java.io.IOException;

public class ChangeCreator {

    private final String gitUrl;
    private Repository repository;

    public ChangeCreator(String gitUrl, FilePath workspace) throws IOException {
        this.gitUrl = gitUrl;
        this.repository = new RepositoryBuilder().setGitDir(
                new File(workspace.getRemote() + "/.git")).build();
    }

    public Change create(ChangeLogSet.Entry changeLogEntry) {
        Author author = find(changeLogEntry.getCommitId());
        if (author == null) {
            author = new Author(
                    changeLogEntry.getAuthor().getId(),
                    AddressSearcher.findById(changeLogEntry.getAuthor().getId()));
        }
        return new Change(
                changeLogEntry.getCommitId().substring(0, 7),
                author,
                changeLogEntry.getMsg(),
                gitUrl.concat("/commit/" + changeLogEntry.getCommitId()));
    }

    private Author find(String commitId) {
        try (RevWalk revWalk = new RevWalk(repository)) {
            RevCommit revCommit = revWalk.parseCommit(ObjectId.fromString(commitId));
            return new Author(revCommit.getAuthorIdent().getName(), revCommit.getAuthorIdent().getEmailAddress());
        } catch (IOException e) {
            return null;
        }
    }

}

File: src/main/kotlin/by/dev/madhead/jarvis/creator/ChangeCreator.kt
package by.dev.madhead.jarvis.creator

import by.dev.madhead.jarvis.model.Author
import by.dev.madhead.jarvis.model.Change
import by.dev.madhead.jarvis.util.AddressSearcher
import hudson.FilePath
import hudson.scm.ChangeLogSet
import org.eclipse.jgit.lib.ObjectId
import org.eclipse.jgit.lib.RepositoryBuilder
import org.eclipse.jgit.revwalk.RevWalk
import java.io.File
import java.io.IOException

class ChangeCreator(
        private val gitUrl: String,
        workspace: FilePath) {

    private val repository = RepositoryBuilder()
            .setGitDir(File(workspace.remote + "/.git")).build()

    fun create(changeLogEntry: ChangeLogSet.Entry): Change {
        return Change(
                revision = changeLogEntry.commitId.substring(0, 7),
                author = findAuthor(changeLogEntry.commitId) ?: Author(
                        changeLogEntry.author.id,
                        AddressSearcher().findById(changeLogEntry.author.id)),
                message = changeLogEntry.msg,
                link = "$gitUrl/commit/${changeLogEntry.commitId}"
        )
    }

    private fun findAuthor(commitId: String): Author? {
        try {
            RevWalk(repository).use {
                val revCommit = it.parseCommit(ObjectId.fromString(commitId))
                return Author(revCommit.authorIdent.name, revCommit.authorIdent.emailAddress)
            }
        } catch (e: IOException) {
            return null
        }
    }

}
